CORDA-3133 (Version 2)

MrMatthewLayton opened the pull request:
### CORDA-3133 (Version 2)

https://r3-cev.atlassian.net/browse/CORDA-3133

This one is a little more flexible by design. This adds an internal `resolveAtTransaction` property to the `StatePointer` class. It's a default property set to `true`, so by default, the behaviour is the same as it is without the fix in place and therefore should be backwards compatible with any existing code. 

This gives developers an opt-in which means that they can choose whether state pointers should be resolved to reference inputs per-state pointer. This does however mean that developers can create output states that point to stale data, simply because it would not be resolved by the transaction builder.

I think the flexibility is worth considering - it puts the control back in the hands of the developer and the restriction can be strict or relaxed on a per-business case, rather than a blanket rule. 

_Apologies for naming: `resolveAtTransaction` actually sounds terrible to me, but it's 7PM and my brain is shutting down!_

MrMatthewLayton committed:
CORDA-3133 [v1]
This proposal is a little more flexible by design. It adds a property to the state pointer class, allowing them to be referenced in transactions on a per-state-pointer basis.

roastario commented:
Can one of the admins verify this patch?

roger3cev commented:
This approach looks good to me! @mnesbit what do you reckon? Thanks for the contribution @series0ne !

roger-that-dev commented on core/src/main/kotlin/net/corda/core/contracts/StatePointer.kt:
```diff
@@ -54,7 +56,11 @@ sealed class StatePointer<T : ContractState> {
+class StaticPointer<T : ContractState>(
+        override val pointer: StateRef,
+        override val type: Class<T>,
+        override val resolveAtTransaction: Boolean = true
```
Maybe `StaticPointer`s should be set to _not_ resolve by default - for the reasons you stated prior.

MrMatthewLayton replied:
@roger3cev I did wonder that, but existing applications that are using state pointers might start to behave unexpectedly when upgraded to a new Corda version, and it wouldn't be immediately apparent as to why.

roger-that-dev replied:
This is true.... But this, in a way, is a bug fix. Will talk with Matthew and get back to you. Cheers

roger-that-dev commented on core/src/main/kotlin/net/corda/core/contracts/StatePointer.kt:
```diff
@@ -14,10 +14,12 @@ import net.corda.core.transactions.LedgerTransaction
 @CordaSerializable
 @KeepForDJVM
-sealed class StatePointer<T : ContractState> {
+sealed class StatePointer<T : ContractState>(internal open val resolveAtTransaction: Boolean = true) {
```
Ya, naming is always tough for stuff like this... What about `isResolved` ?

mnesbit replied:
This probably shouldn't be internal because name mangling and serialization will not interact at all well. Is it ok to make it private? It can always have a public read only val.

shamsasari replied:
It can't be private as it's needed by TransactionBuilder

MrMatthewLayton replied:
My rationale was that this probably doesn't need to be seen by consuming code beyond the constructor, except for in the transaction builder where it's needed to determine whether or not to resolve the state pointer to a reference input.

As another thought, you could make it public instead. This would mean that you could use it in your contract code if needed; for example:

```
"Resolved pointers must point to the specified reference input" using (
    if(stateWithPointer.resolveAtTransaction) {
        // make checks against your reference input
    }
)
```

roger-that-dev replied:
This has been addressed

shamsasari commented on core/src/main/kotlin/net/corda/core/contracts/StatePointer.kt:
```diff
@@ -14,10 +14,12 @@ import net.corda.core.transactions.LedgerTransaction
 @CordaSerializable
 @KeepForDJVM
-sealed class StatePointer<T : ContractState> {
+sealed class StatePointer<T : ContractState>(internal open val resolveAtTransaction: Boolean = true) {
```
In line with the other properties, can you declare this as an abstract property below (where `pointer` and `type` are) and not give it a default value. The default value is not necessary here as the concrete classes below specify them.

Also any reason why you made this internal? I think it's OK to have this public, the subclasses make this public anyway.

roger-that-dev replied:
This has been addressed

shamsasari commented on core/src/main/kotlin/net/corda/core/contracts/StatePointer.kt:
```diff
@@ -54,7 +56,11 @@ sealed class StatePointer<T : ContractState> {
+class StaticPointer<T : ContractState>(
+        override val pointer: StateRef,
+        override val type: Class<T>,
+        override val resolveAtTransaction: Boolean = true
```
You will need to annotate the new c'tor with `@JvmOverloads` to preserve the old one that's now part of the API. This is probably why the API check is failing.

shamsasari commented on core/src/main/kotlin/net/corda/core/contracts/StatePointer.kt:
```diff
@@ -99,7 +105,11 @@ class StaticPointer<T : ContractState>(override val pointer: StateRef, override
+class LinearPointer<T : LinearState>(
+        override val pointer: UniqueIdentifier,
+        override val type: Class<T>,
+        override val resolveAtTransaction: Boolean = true
```
ditto

roger3cev commented:
This is true.... But this, in a way, is a bug fix. Will talk with Matthew and get back to you. Cheers

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Matthew Layton <notifications@github.com>
Sent: Thursday, September 5, 2019 10:23:25 AM
To: corda/corda <corda@noreply.github.com>
Cc: Roger Willis <roger.willis@r3.com>; Mention <mention@noreply.github.com>
Subject: Re: [corda/corda] CORDA-3133 (Version 2) (#5428)


@series0ne commented on this pull request.

________________________________

In core/src/main/kotlin/net/corda/core/contracts/StatePointer.kt<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fcorda%2Fcorda%2Fpull%2F5428%23discussion_r321158862&data=02%7C01%7C%7Cf063b7d9ec9d4f01691308d731e2b54a%7Ca4be1f2e2d10419587cd736aca9b672c%7C0%7C0%7C637032722087919414&sdata=yxCqhnwBT3VZYoXkEgVD3p4o99%2FZ4s8h%2BlGoYqPQzOc%3D&reserved=0>:

> @@ -54,7 +56,11 @@ sealed class StatePointer<T : ContractState> {
  *   throw a [TransactionResolutionException]
  */
 @KeepForDJVM
-class StaticPointer<T : ContractState>(override val pointer: StateRef, override val type: Class<T>) : StatePointer<T>() {
+class StaticPointer<T : ContractState>(
+        override val pointer: StateRef,
+        override val type: Class<T>,
+        override val resolveAtTransaction: Boolean = true


@roger3cev<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Froger3cev&data=02%7C01%7C%7Cf063b7d9ec9d4f01691308d731e2b54a%7Ca4be1f2e2d10419587cd736aca9b672c%7C0%7C0%7C637032722087919414&sdata=1uUv3%2Be6L%2B6Q27y6TAX6Ehr1COIOnrCXLzcvTDw3e2Q%3D&reserved=0> I did wonder that, but existing applications that are using state pointers might start to behave unexpectedly when upgraded to a new Corda version, and it wouldn't be immediately apparent as to why.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fcorda%2Fcorda%2Fpull%2F5428%3Femail_source%3Dnotifications%26email_token%3DAFMSO7SYI2CUNP7ZDX2TEKTQIDFY3A5CNFSM4ITJWKVKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCDXWNYQ%23discussion_r321158862&data=02%7C01%7C%7Cf063b7d9ec9d4f01691308d731e2b54a%7Ca4be1f2e2d10419587cd736aca9b672c%7C0%7C0%7C637032722087929410&sdata=r%2BHEBjwru6hO4uQP%2FMOAO6StpkMYrr5XQceETihzL%2BA%3D&reserved=0>, or mute the thread<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAFMSO7SN3ZPGZAY5NXNIF23QIDFY3ANCNFSM4ITJWKVA&data=02%7C01%7C%7Cf063b7d9ec9d4f01691308d731e2b54a%7Ca4be1f2e2d10419587cd736aca9b672c%7C0%7C0%7C637032722087939407&sdata=u2LhzwsKh0KmOX6Gr2NsQQNwIMStr1Kko6Tyym23m50%3D&reserved=0>.

shamsasari commented on core/src/main/kotlin/net/corda/core/contracts/StatePointer.kt:
```diff
@@ -54,7 +56,11 @@ sealed class StatePointer<T : ContractState> {
+class StaticPointer<T : ContractState>(
+        override val pointer: StateRef,
+        override val type: Class<T>,
+        override val resolveAtTransaction: Boolean = true
```
Also, I just realised that `StatePointer` is corda serialisable, which means we need to preserve compatibility with the old serialisation format that doesn't include `resolveAtTransaction`. Normally this is all done seamlessly if we're only adding new properties which are nullable. `resolveAtTransaction` is not so unfortunately we need to add a bit more boilerplate. It's all explained here https://docs.corda.net/serialization-default-evolution.html.

roger-that-dev replied:
Ah yes good spot!

roger-that-dev replied:
Looks like this is resolved now

MrMatthewLayton committed:
CORDA-3133
 - Updated `resolveAtTransaction` to `isResolved`.
 - Moved `isResolved` out of the constructor into an abstract property.
 - Added deprecation constructor declaration for backwards compatibility.

series0ne commented:
I've added some updates to this PR. For some reason it's not building on my machine due to javafx dependencies, though this seems like more of a local issue if anything.

If there are any further issues with this, it might be worth me coming into R3 and sitting down with someone to get this over the line.

roger-that-dev commented on core/src/main/kotlin/net/corda/core/contracts/StatePointer.kt:
```diff
@@ -54,7 +60,21 @@ sealed class StatePointer<T : ContractState> {
  */
 @KeepForDJVM
-class StaticPointer<T : ContractState>(override val pointer: StateRef, override val type: Class<T>) : StatePointer<T>() {
+class StaticPointer<T : ContractState>(
```
Add `JvmOverloads` here

shamsasari replied:
Actually it's not needed as the other overloaded c'tor is defined below.

roger-that-dev replied:
Ah gosh - ignore me @series0ne :) thanks @shamsasari

roger-that-dev commented on core/src/main/kotlin/net/corda/core/contracts/StatePointer.kt:
```diff
@@ -99,7 +119,21 @@ class StaticPointer<T : ContractState>(override val pointer: StateRef, override
  */
 @KeepForDJVM
-class LinearPointer<T : LinearState>(override val pointer: UniqueIdentifier, override val type: Class<T>) : StatePointer<T>() {
+class LinearPointer<T : LinearState>(
```
Add JvmOverloads here

shamsasari replied:
likewise

shamsasari commented on core/src/main/kotlin/net/corda/core/contracts/StatePointer.kt:
```diff
@@ -28,6 +29,11 @@ sealed class StatePointer<T : ContractState> {
+    /**
+     * Determines whether the state pointer should be resolved to a reference input when used in a transaction.
+     */
+    abstract val isResolved: Boolean
```
The API checker is still complaining. It's because we've added a new abstract method but the class is not annotated with `DoNotImplement`.

The easiest way to resolve this is to make it non-abstract and make it return the default value.
`open val isResolved: Boolean: get() = false`

There's a separate discussion to be had on whether we should annotate `StatePointer` with `@DoNotImplement`. It seems to be an omission that we didn't annotate it in the first place. Perhaps we should do it before the next release.

roger3cev commented:
@shamsasari Yes, it makes sense to annotate it with `DoNotImplement`

roger-that-dev reviewed (commented):
Looks like it's nearly there - just need to get the API checker to pass. See @shamsasari 's comments. Ping me on slack if you get stuck. Cheers!

MrMatthewLayton committed:
CORDA-3133 (version 2)
 - Added required changes to state pointers as per PR comments.
 - Added unit tests to ensure isResolved can be used to configure when state pointers should be resolved to reference inputs.

roger3cev commented:
Boom API checks passing @series0ne !

shamsasari commented on core/src/main/kotlin/net/corda/core/contracts/StatePointer.kt:
```diff
@@ -17,6 +19,7 @@ import net.corda.core.transactions.LedgerTransaction
  */
 @CordaSerializable
 @KeepForDJVM
+@DoNotImplement
```
Can you add an entry to the changelog.rst saying that `StatePointer` has been marked as `@DoNotImplement`, which was an omission in the original release.

MrMatthewLayton committed:
CORDA-3133 (version 2)
 - fixed unit tests.
 - added comment to changelog.rst.
 - added helper functions to `StatePointer` to allow easier creation of static and linear pointers.

Commit:
Merge branch 'release/os/4.3' into CORDA-3133-v2

shamsasari commented on core/src/main/kotlin/net/corda/core/contracts/StatePointer.kt:
```diff
@@ -17,7 +19,36 @@ import net.corda.core.transactions.LedgerTransaction
+         * @param isResolved Determines whether the state pointer should be resolved to a reference input when included in a transaction.
+         * @return Returns a [StaticPointer] to the specified contract state.
+         */
+        inline fun <reified T : ContractState> staticPointer(
```
Hmm, I understand the reason for having this but I'm not sure this is the PR to do it in. I'd rather this PR just deal with CORDA-3133. Feel free to raise a separate PR (you will have to check that it makes sense in Java code).

roger-that-dev replied:
I said it's ok for him to merge in the other PR - it was approved but floundering as CI kept on takes ages then it needed a rebase etc...

shamsasari replied:
Fair enough.

roastario commented:
Build
